pkg/kv/kvserver/kvstorage/init.go

```diff
@@ -291,103 +293,198 @@ func IterateRangeDescriptorsFromDisk(
-			// was ever written. So we backfill the replicaID here (as of 23.1) and
-			// remove this code in the future (the follow-up release, assuming it is
-			// forced to migrate through 23.1, otherwise later).
-			if buildutil.CrdbTestBuild {
-				return nil, errors.AssertionFailedf("%s has no persisted replicaID", initM[id])
+	// Check invariants. Sorted order for deterministic unit tests.
+	//
+	// Migrate into RaftReplicaID for all replicas that need it.
+	for idx, repl := range sl {
+		var descReplicaID roachpb.ReplicaID
+		if repl.Desc != nil {
+			// INVARIANT: a Replica's RangeDescriptor always contains the local Store,
+			// i.e. a Store is a member of all of its local initialized Replicas.
+			replDesc, found := repl.Desc.GetReplicaDescriptor(ident.StoreID)
+			if !found {
+				return nil, errors.AssertionFailedf("s%d not found in %s", ident.StoreID, repl.Desc)
 			}
-			if err := logstore.NewStateLoader(id.RangeID).SetRaftReplicaID(ctx, eng, id.ReplicaID); err != nil {
-				return nil, errors.Wrapf(err, "backfilling replicaID for r%d", id.RangeID)
+			if repl.ReplicaID != 0 && replDesc.ReplicaID != repl.ReplicaID {
+				return nil, errors.AssertionFailedf("conflicting RaftReplicaID %d for %s", repl.ReplicaID, repl.Desc)
+			}
+			descReplicaID = replDesc.ReplicaID
+		}
+
+		if repl.ReplicaID != 0 {
+			// RaftReplicaID present, no need to migrate.
+			continue
+		}
+
+		// Migrate into RaftReplicaID. This migration can be removed once the
+		// BinaryMinSupportedVersion is >= 23.1, and we can assert that
+		// repl.ReplicaID != 0 always holds.
+
+		if descReplicaID != 0 {
+			// Backfill RaftReplicaID for an initialized Replica.
+			if err := logstore.NewStateLoader(repl.RangeID).SetRaftReplicaID(ctx, eng, descReplicaID); err != nil {
+				return nil, errors.Wrapf(err, "backfilling ReplicaID for r%d", repl.RangeID)
+			}
+			repl.ReplicaID = descReplicaID
+			sl[idx] = repl
+			log.Eventf(ctx, "backfilled replicaID for initialized replica %s", repl.ID())
+		} else {
+			// We found an uninitialized replica that did not have a persisted
+			// ReplicaID. We can't determine the ReplicaID now, so we migrate by
+			// removing this uninitialized replica. This technically violates raft
+			// invariants if this replica has cast a vote, but the conditions under
+			// which this matters are extremely unlikely.
+			//
+			// TODO(tbg): if clearRangeData were in this package we could destroy more
+			// effectively even if for some reason we had in the past written state
+			// other than the HardState here (not supposed to happen, but still).
+			if err := eng.ClearUnversioned(logstore.NewStateLoader(repl.RangeID).RaftHardStateKey()); err != nil {
+				return nil, errors.Wrapf(err, "removing HardState for r%d", repl.RangeID)
 			}
-			log.Eventf(ctx, "backfilled replicaID for %s", id)
+			log.Eventf(ctx, "removed legacy uninitialized replica for r%s", repl.RangeID)
+			sl = append(sl[:idx], sl[idx+1:]...)
```

Comment:
To avoid the O(N * M) complexity here, in a case when N out of M replicas are legacy uninited replicas, could instead rewrite one `Replica` in the slice at a time:

```golang
back := 0
for idx, repl := range sl {
	...
	if !(remove condition) {
		sl[back] = repl
		back++
	}
}
return sl[:back], nil
```

May require a bit of shuffling/reordering to work around the `continue`.